### PR TITLE
Allow passing in cr/cl bounds and other settings

### PR DESCRIPTION
**Pull request by winston-zillow**
Fix execution on CPU and GPU. Fix model loading.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -22,7 +22,7 @@ We need to put the data sets in the `dataset` folder. You can specify one data s
 
 ```bash
 # trained on the tic-tac-toe data set with one GPU.
-python3 experiment.py -d tic-tac-toe -bs 32 -s 1@16 -e401 -lrde 200 -lr 0.002 -ki 0 -mp 12481 -i 0 -wd 1e-6 &
+python3 experiment.py -d tic-tac-toe -bs 32 -s 1@16 -e401 -lrde 200 -lr 0.002 -ki 0 -mp 12481 -i cuda:0 -wd 1e-6 &
```
Comment (winston-zillow) [inline README.md:25]: Note: see review comment on `args.py` changes

File: args.py [hunk 2]
```diff
@@ -51,7 +52,8 @@
 rrl_args.plot_file = os.path.join(rrl_args.folder_path, 'plot_file.pdf')
 rrl_args.log = os.path.join(rrl_args.folder_path, 'log.txt')
 rrl_args.test_res = os.path.join(rrl_args.folder_path, 'test_res.txt')
-rrl_args.device_ids = list(map(int, rrl_args.device_ids.strip().split('@')))
+rrl_args.device_ids = list(map(lambda id: torch.device(id), rrl_args.device_ids.strip().split('@'))) \
+    if rrl_args.device_ids else [None]
```
Comment (winston-zillow) [inline args.py:56]: Note: I found that passing in integer device ID would get the tensors pegged to the GPU memory but the GPU compute utilization remains at 0, as shown by `nvidia-smi`. After I change the device ID to that returned by `torch.device("cuda:0")`, the GPU is utilized fully. I do not know why that's the case as simple test using a python loop can cause GPU utilization.

Example run passing in integer device ID:
```
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 450.142.00   Driver Version: 450.142.00   CUDA Version: 11.0     |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|                               |                      |               MIG M. |
|===============================+======================+======================|
|   0  Tesla K80           On   | 00000000:00:1E.0 Off |                    0 |
| N/A   47C    P0    70W / 149W |    322MiB / 11441MiB |      0%      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+
                                                                               
+-----------------------------------------------------------------------------+
| Processes:                                                                  |
|  GPU   GI   CI        PID   Type   Process name                  GPU Memory |
|        ID   ID                                                   Usage      |
|=============================================================================|
|    0   N/A  N/A     27173      C   ...vs/pytorch_p37/bin/python      319MiB |
+-----------------------------------------------------------------------------+
```

Example run passing in `cuda:*`:

```
+-----------------------------------------------------------------------------+
| Processes:                                                                  |
|  GPU   GI   CI        PID   Type   Process name                  GPU Memory |
|        ID   ID                                                   Usage      |
|=============================================================================|
|    0   N/A  N/A     27346      C   ...vs/pytorch_p37/bin/python     1736MiB |
+-----------------------------------------------------------------------------+
Sat Dec  4 01:31:31 2021       
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 450.142.00   Driver Version: 450.142.00   CUDA Version: 11.0     |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|                               |                      |               MIG M. |
|===============================+======================+======================|
|   0  Tesla K80           On   | 00000000:00:1E.0 Off |                    0 |
| N/A   52C    P0   138W / 149W |   1739MiB / 11441MiB |    100%      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+
                                                                               
+-----------------------------------------------------------------------------+
| Processes:                                                                  |
|  GPU   GI   CI        PID   Type   Process name                  GPU Memory |
|        ID   ID                                                   Usage      |
|=============================================================================|
|    0   N/A  N/A     27346      C   ...vs/pytorch_p37/bin/python     1736MiB |
+-----------------------------------------------------------------------------+
```

File: experiment.py [hunk 3]
```diff
@@ -14,15 +14,37 @@
...
+            'normalize_continuous': True,
+            'one_hot_encode_features': True,
+            'impute_continuous': True,
+            # of shape [continious columns, lower bounds, upper bounds]
+            'bounds': None
+            # alternatively, pass in individual bounds
+            # lower_bound: [continuous cols]
+            # upper_bound: [continuous cols]
+        }
+    return settings
```
Comment (winston-zillow) [inline experiment.py:32]: Note: I added this new setting file so that the user can pass in CR/CL bounds as well as controlling normalization and one-hot encoding etc. (those are currently hard-coded)

File: rrl/components.py [hunk 4]
```diff
@@ -39,12 +39,17 @@ def __init__(self, n, input_dim, use_not=False, left=None, right=None):
         self.register_buffer('right', right)
 
         if self.input_dim[1] > 0:
-            if self.left is not None and self.right is not None:
+            if cl is not None and cr is not None:  # bounds are specified
+                cl = torch.tensor(cl).type(torch.float).t()
+                cr = torch.tensor(cr).type(torch.float).t()
```
Comment (winston-zillow) [inline rrl/components.py:44]: Note: here we can pass in the cl/cr bounds directly.

File: rrl/components.py [hunk 5]
```diff
@@ -39,12 +39,17 @@ def __init__(self, n, input_dim, use_not=False, left=None, right=None):
...
+            elif self.left is not None and self.right is not None:
                 cl = self.left + torch.rand(self.n, self.input_dim[1]) * (self.right - self.left)
                 cr = self.left + torch.rand(self.n, self.input_dim[1]) * (self.right - self.left)
             else:
                 cl = 3. * (2. * torch.rand(self.n, self.input_dim[1]) - 1.)
                 cr = 3. * (2. * torch.rand(self.n, self.input_dim[1]) - 1.)
+            assert torch.Size([self.n, self.input_dim[1]]) == cl.size()
+            assert torch.Size([self.n, self.input_dim[1]]) == cr.size()
```
Comment (winston-zillow) [inline rrl/components.py:52]: Note: and verify the shapes are correct.

File: rrl/models.py [hunk 6]
```diff
@@ -99,9 +99,11 @@ def __init__(self, dim_list, device_id, use_not=False, is_rank0=False, log_file=
                 logging.basicConfig(level=logging.DEBUG, filename=log_file, filemode='w', format=log_format)
         self.writer = writer
 
-        self.net = MLLP(dim_list, use_not=use_not, left=left, right=right, estimated_grad=estimated_grad)
+        self.net = MLLP(dim_list, use_not=use_not, cl=cl, cr=cr, left=left, right=right,
+                        estimated_grad=estimated_grad)
 
-        self.net.cuda(self.device_id)
+        if self.device_id and self.device_id.type == 'cuda':
```
Comment (winston-zillow) [inline rrl/models.py:105]: Note: the condition allows the program to run in CPU mode as well.

File: rrl/utils.py [hunk 7]
```diff
@@ -29,13 +30,14 @@ def read_csv(data_path, info_path, shuffle=False):
...
+    def __init__(self, f_df, discrete=False, y_one_hot=True, drop='first',
+                 impute_continuous=True, one_hot_encode_features=True):
         self.f_df = f_df
         self.discrete = discrete
         self.y_one_hot = y_one_hot
         self.label_enc = preprocessing.OneHotEncoder(categories='auto') if y_one_hot else preprocessing.LabelEncoder()
-        self.feature_enc = preprocessing.OneHotEncoder(categories='auto', drop=drop)
-        self.imp = SimpleImputer(missing_values=np.nan, strategy='mean')
+        self.feature_enc = preprocessing.OneHotEncoder(categories='auto', drop=drop) if one_hot_encode_features else None
+        self.imp = SimpleImputer(missing_values=np.nan, strategy='mean') if impute_continuous else None
```
Comment (winston-zillow) [inline rrl/utils.py:40]: Note: for dataset not requiring or already have one-hot encoding or imputation, they can now be skipped.

**Comment by 12wang3** [comment]
Thank you very much for the PR. I am busy on other stuff now and will check the code after Dec 9.

**Comment by ASan1527** [comment]
image
I cant catch the device_ids, and I only have the single gpu, I don't know how to change the code. Could you please tell me to solve it? thank you!

**Comment by 12wang3** [comment]
> image I cant catch the device_ids, and I only have the single gpu, I don't know how to change the code. Could you please tell me to solve it? thank you!

Could you please show the command you used? Have you set the "-i" argument? It seems that you did not set the device_ids since your device_ids was None. If you only have one single GPU, you can use "-i 0" to set the device_ids. By the way, maybe we should use issue rather than PR to discuss questions.
